File: app/drafting/drafter.py
from datetime import datetime

from app.drafting.prompts import build_instruction_block
from app.extraction.schemas import ResolutionData


def _render_section(title: str, lines: list[str]) -> str:
    if not lines:
        return f"{title}\n- Not provided\n"
    return f"{title}\n" + "\n".join(f"- {line}" for line in lines) + "\n"


def draft_resolution(resolution_data: ResolutionData, instructions: str) -> str:
    instruction_block = build_instruction_block(instructions)
    now = datetime.utcnow().strftime("%Y-%m-%d")

    header = (
        f"Draft Legal Memorandum/Resolution\n"
        f"Date: {now}\n"
        f"Document Type: {resolution_data.document_type}\n"
        f"Company: {resolution_data.company_name or 'TBD'}\n"
    )

    body_sections = [
        _render_section("Registered Office", [resolution_data.registered_office]),
        _render_section("Key Dates", resolution_data.key_dates),
        _render_section("Directors", resolution_data.directors),
        _render_section("Shareholders", resolution_data.shareholders),
        _render_section("Share Capital", [resolution_data.share_capital]),
        _render_section("Objectives/Objects", resolution_data.objectives),
        _render_section("Key Clauses", resolution_data.key_clauses),
        _render_section("Governance Notes", resolution_data.governance_notes),
        _render_section("Legal Authority References", resolution_data.authorities),
        _render_section(
            "Parties",
            [f"{party.name} ({party.role})" for party in resolution_data.parties],
        ),
        _render_section("Warnings", resolution_data.warnings),
    ]

    draft = (
        f"{header}\n"
        f"Instructions Applied:\n{instruction_block}\n\n"
        + "\n".join(body_sections)
        + "\nPrepared by the Legal AI Drafting Assistant."
    )

    return draft

File: app/web/src/core/drafter.js
const formatHeader = (result, format) => {
  const today = new Date().toISOString().split("T")[0];
  const formatLabel = format === "memo" ? "Memo" : format === "brief" ? "Brief" : "Resolution";
  return [
    `Draft ${formatLabel}`,
    `Date: ${today}`,
    `Document Type: ${result.documentType}`,
    `Company: ${result.companyName || "TBD"}`,
  ].join("\n");
};

const renderSection = (title, lines) => {
  if (!lines || lines.length === 0) {
    return `${title}\n- Not provided`;
  }
  const formatted = Array.isArray(lines) ? lines : [lines];
  return `${title}\n${formatted.map((line) => `- ${line}`).join("\n")}`;
};

const buildChecklist = (result) => [
  "Confirm board approval form (written resolution vs meeting minutes).",
  "Verify registered office address matches filings.",
  "Check quorum and notice requirements referenced in Articles.",
  result.authorities.length
    ? "Cross-check relevant Companies Act sections cited."
    : "Confirm legal authority citations with counsel.",
];

const buildRiskNotes = (result) => {
  const risks = [];
  if (!result.companyName) risks.push("Company name missing.");
  if (!result.registeredOffice) risks.push("Registered office not detected.");
  if (!result.keyClauses.length) risks.push("Key clauses not detected.");
  if (!result.actionItems.length) risks.push("No action items detected in correspondence.");
  return risks.length ? risks : ["No critical gaps detected from the extracted text."];
};

export const buildDraft = ({
  result,
  instructions,
  format,
  includeRisk,
  includeChecklist,
  includeTimeline,
  includeClauseMap,
}) => {
  const sections = [
    renderSection("Registered Office", result.registeredOffice || ""),
    renderSection("Key Dates", result.keyDates),
    renderSection("Directors", result.directors),
    renderSection("Share Capital", result.shareCapital || ""),
    renderSection("Objectives", result.objectives),
    renderSection("Key Clauses", result.keyClauses),
    renderSection("Governance Notes", result.governanceNotes),
    renderSection(
      "Parties",
      result.parties.map((party) => `${party.name} (${party.role})`)
    ),
    renderSection("Action Items", result.actionItems),
    renderSection("Authorities", result.authorities),
  ];

  if (includeTimeline) {
    sections.push(
      renderSection(
        "Timeline",
        result.timeline.map((item) => `${item.date} — ${item.summary}`)
      )
    );
  }

  if (includeClauseMap) {
    sections.push(
      renderSection(
        "Clause Map",
        result.clauseMap.map((item) => `${item.label} — ${item.summary}`)
      )
    );
  }

  if (includeRisk) {
    sections.push(renderSection("Risk Flags", buildRiskNotes(result)));
  }

  if (includeChecklist) {
    sections.push(renderSection("Filing Checklist", buildChecklist(result)));
  }

  return [
    formatHeader(result, format),
    "",
    "Instructions:",
    instructions || "None provided.",
    "",
    sections.join("\n\n"),
    "",
    "Prepared by the Legal AI Drafting Studio.",
  ].join("\n");
};
